From 48dc51237fc2884d14b47a05cbdfdc8ca764b083 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Tue, 30 Dec 2025 14:50:36 +0800
Subject: [PATCH 1/7] mtk: mt76: mt7996: fix MT7992 precal layout issue

Align the layout of MT7992 precal with logan
In order to be compatible with MT7992 2+6 sku, the layout reserves the
partition of 6G group precal

Also, add wait_on_prek_offs in group precal, as the firmware return size
is not same as group_size.
Firmware returns 2+5G data or 2+6G data with a size of GROUP_SIZE_2G +
max(GROUP_SIZE_5G, GROUP_SIZE_6G)

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7996/eeprom.h   | 11 +++++++----
 mt7996/testmode.c | 29 ++++++++++++++++++++++++-----
 2 files changed, 31 insertions(+), 9 deletions(-)

diff --git a/mt7996/eeprom.h b/mt7996/eeprom.h
index d2b88164a..344aa66db 100644
--- a/mt7996/eeprom.h
+++ b/mt7996/eeprom.h
@@ -54,6 +54,8 @@ enum mt7996_prek_rev {
 	GROUP_SIZE_2G,
 	GROUP_SIZE_5G,
 	GROUP_SIZE_6G,
+	GROUP_SIZE_2G_V2,
+	GROUP_SIZE_ABAND_V2,
 	ADCDCOC_SIZE_2G,
 	ADCDCOC_SIZE_5G,
 	ADCDCOC_SIZE_6G,
@@ -86,14 +88,13 @@ static const u32 mt7996_prek_rev_233[] = {
 	[DPD_OTFG0_SIZE] =			2 * MT_EE_CAL_UNIT,
 };
 
-/* kite 2/5g config */
 static const u32 mt7992_prek_rev[] = {
 	[GROUP_SIZE_2G] =			4 * MT_EE_CAL_UNIT,
-	[GROUP_SIZE_5G] =			110 * MT_EE_CAL_UNIT,
-	[GROUP_SIZE_6G] =			0,
+	[GROUP_SIZE_5G] =			45 * MT_EE_CAL_UNIT,
+	[GROUP_SIZE_6G] =			110 * MT_EE_CAL_UNIT,
 	[ADCDCOC_SIZE_2G] =			4 * 4,
 	[ADCDCOC_SIZE_5G] =			4 * 5,
-	[ADCDCOC_SIZE_6G] =			0,
+	[ADCDCOC_SIZE_6G] =			4 * 5,
 	[DPD_LEGACY_SIZE] =			5 * MT_EE_CAL_UNIT,
 	[DPD_MEM_SIZE] =			16 * MT_EE_CAL_UNIT,
 	[DPD_OTFG0_SIZE] =			2 * MT_EE_CAL_UNIT,
@@ -103,6 +104,8 @@ static const u32 mt7990_prek_rev[] = {
 	[GROUP_SIZE_2G] =			4 * MT_EE_CAL_UNIT,
 	[GROUP_SIZE_5G] =			45 * MT_EE_CAL_UNIT,
 	[GROUP_SIZE_6G] =			90 * MT_EE_CAL_UNIT,
+	[GROUP_SIZE_2G_V2] =			6 * MT_EE_CAL_UNIT,
+	[GROUP_SIZE_ABAND_V2] =			106 * MT_EE_CAL_UNIT,
 	[ADCDCOC_SIZE_2G] =			4 * 2,
 	[ADCDCOC_SIZE_5G] =			4 * 3,
 	[ADCDCOC_SIZE_6G] =			4 * 3,
diff --git a/mt7996/testmode.c b/mt7996/testmode.c
index f3c0b7692..3840efc58 100644
--- a/mt7996/testmode.c
+++ b/mt7996/testmode.c
@@ -532,7 +532,8 @@ mt7996_tm_group_prek(struct mt7996_phy *phy, enum mt76_testmode_state state)
 		},
 	};
 	u32 i, group_size, dpd_size, size, offs, *pre_cal;
-	u8 *eeprom, do_precal;
+	u32 wait_on_prek_offset;
+	u8 *eeprom, do_precal = 0;
 	int ret = 0;
 
 	if (!dev->flash_mode) {
@@ -547,9 +548,26 @@ mt7996_tm_group_prek(struct mt7996_phy *phy, enum mt76_testmode_state state)
 	dpd_size = MT_EE_CAL_DPD_SIZE;
 	size = group_size + dpd_size;
 	offs = MT_EE_DO_PRE_CAL;
-	do_precal = (MT_EE_WIFI_CAL_GROUP_2G * !!PREK(GROUP_SIZE_2G)) |
-		    (MT_EE_WIFI_CAL_GROUP_5G * !!PREK(GROUP_SIZE_5G)) |
-		    (MT_EE_WIFI_CAL_GROUP_6G * !!PREK(GROUP_SIZE_6G));
+
+	if (mt7996_band_phy(dev, NL80211_BAND_2GHZ))
+		do_precal |= MT_EE_WIFI_CAL_GROUP_2G;
+	if (mt7996_band_phy(dev, NL80211_BAND_5GHZ))
+		do_precal |= MT_EE_WIFI_CAL_GROUP_5G;
+	if (mt7996_band_phy(dev, NL80211_BAND_6GHZ))
+		do_precal |= MT_EE_WIFI_CAL_GROUP_6G;
+
+	/* For 7990, the flash memory layout uses the old definitions,
+	 * but the valid size follows the new one
+	 */
+	if (is_mt7990(&dev->mt76))
+		wait_on_prek_offset = PREK(GROUP_SIZE_2G_V2) + PREK(ADCDCOC_SIZE_2G) +
+				      PREK(GROUP_SIZE_ABAND_V2) + PREK(ADCDCOC_SIZE_5G);
+	else if (is_mt7992(&dev->mt76))
+		wait_on_prek_offset = PREK(GROUP_SIZE_2G) + PREK(ADCDCOC_SIZE_2G) +
+				      max(PREK(GROUP_SIZE_5G) + PREK(ADCDCOC_SIZE_5G),
+					  PREK(GROUP_SIZE_6G) + PREK(ADCDCOC_SIZE_6G));
+	else
+		wait_on_prek_offset = group_size;
 
 	switch (state) {
 	case MT76_TM_STATE_GROUP_PREK:
@@ -561,7 +579,8 @@ mt7996_tm_group_prek(struct mt7996_phy *phy, enum mt76_testmode_state state)
 
 		ret = mt76_mcu_send_msg(&dev->mt76, MCU_WM_UNI_CMD(TESTMODE_CTRL), &req,
 					sizeof(req), false);
-		wait_event_timeout(mdev->mcu.wait, dev->cur_prek_offset == group_size,
+		wait_event_timeout(mdev->mcu.wait,
+				   dev->cur_prek_offset == wait_on_prek_offset,
 				   30 * HZ);
 
 		if (ret)
-- 
2.45.2

